app/src/androidTest/java/com/github/polypoly/app/commons/MockDB.kt

```diff
@@ -0,0 +1,74 @@
+package com.github.polypoly.app.commons
+
+import com.github.polypoly.app.network.IRemoteStorage
+import java.util.concurrent.CompletableFuture
+import kotlin.reflect.KClass
+
+class MockDB: IRemoteStorage {
+
+    private val keysHierarchy: MutableMap<String, MutableList<String>> = mutableMapOf()
+    private val data: MutableMap<String, Any> = mutableMapOf()
+
+    private fun cleanKey(key: String): String {
+        return key.removePrefix("/").removeSuffix("/")
+    }
+
+    override fun <T : Any> getValue(key: String, clazz: KClass<T>): CompletableFuture<T> {
+        val keyCleaned = cleanKey(key)
+        if (!data.containsKey(keyCleaned))
+            return CompletableFuture.failedFuture(IllegalAccessException("Invalid key $keyCleaned"))
+        @Suppress("UNCHECKED_CAST")
+        return CompletableFuture.completedFuture(data[cleanKey(key)] as T)
+    }
+
+    override fun <T : Any> getAllValues(key: String, clazz: KClass<T>): CompletableFuture<List<T>> {
+        val objs = mutableListOf<T>()
+        val parentKeyCleaned = cleanKey(key)
+        @Suppress("UNCHECKED_CAST")
+        for (child in keysHierarchy[parentKeyCleaned] ?: listOf())
+            objs.add(data["$parentKeyCleaned/$child"] as T)
```

Comment:
You can use safe cast instead :
`(data["$parentKeyCleaned/$child" as? T)?.let { objs.add(it) }` (i.e. if the cast succeeds and the key exist, add it, else do nothing)

Reply:
This was actually an intended feature: this class is designed for test purposes, not for live application deployment. Therefore, if somehow an incorrect value is put in the DB which results in a failure when retrieving it (either through `getValue` or `getAllValues`), this means either the test or the code tested had issues that led to an incorrect DB state, so it should crash to reflect this situation (through the exception thrown by the unsafe cast).

Reply:
ok I understand!